Allow customers to assert facts about the current transaction.

eoger opened the pull request:
Fixes #225

rnewman commented on db/src/tx.rs:
```diff
@@ -616,6 +624,17 @@ impl<'conn, 'a> Tx<'conn, 'a> {
                     }
 
+                    if e == KnownEntid(self.tx_id) && a == entids::DB_TX_INSTANT {
+                        tx_instant_set = true;
```
Consider whether it is an error to have:

```
[:db/tx :db/txInstant #inst "2017-01-01T10:00:00Z"]
[:db/tx :db/txInstant #inst "2017-01-01T22:00:00Z"]
```

rnewman commented on db/src/tx.rs:
```diff
@@ -616,6 +624,17 @@ impl<'conn, 'a> Tx<'conn, 'a> {
+                            panic!("This type error should have been caught earlier.");
+                        }
+                        // TODO: INSTANT to be strictly after the last transaction
+                        // timestamp and strictly before the current transactor timestamp.
```
We might want to weaken this condition — Datomic can assume correct clocks, but embedded clients cannot.

And we will need a way to bypass this, too, for syncing.

rnewman commented on db/src/tx.rs:
```diff
@@ -628,13 +647,14 @@ impl<'conn, 'a> Tx<'conn, 'a> {
+            non_fts_one.push((self.tx_id,
+                              entids::DB_TX_INSTANT,
+                              self.schema.require_attribute_for_entid(entids::DB_TX_INSTANT).unwrap(),
+                              TypedValue::Instant(self.tx_instant),
```
So this would be:

```
self.tx_instant.map_or_else(TypedValue::current_instant, TypedValue::Instant)
```

rnewman commented on tx/src/entities.rs:
```diff
@@ -23,20 +23,21 @@ use self::edn::symbols::NamespacedKeyword;
         match self {
             TempId::Internal(x) => Some(x),
-            TempId::External(_) => None,
+            _ => None,
```
The trick you used to find these locations — add the `enum` case, look for the compile errors — stops working if you use `_`. Add the explicit `TempId::Tx` case here instead:

```
TempId::Internal(x) => Some(x),
TempId::Tx |
TempId::External(_) => None,
```

eoger commented:
Thank you @rnewman, I have amended my patch.
Here is what I came up with regarding multiple values passed to txInstant:

```
[:db/add :db/tx :db/txInstant A]
[:db/add :db/tx :db/txInstant B]
// The last :db/add value is always picked up. Here txInstant = B.
```

```
[:db/add :db/tx :db/txInstant A]
[:db/retract :db/tx :db/txInstant A]
// If the last assertion about txInstant is a :db/retract and it corresponds to the last :db/add value, the system picks up a value. Here txInstant = ::now()

[:db/add :db/tx :db/txInstant A]
[:db/retract :db/tx :db/txInstant B]
// Here, the system would give txInstant = A
```

rnewman commented:
N.B., retractions need to pass through; they conceptually apply before the additions.

After working on #533 I think duplicate specifications within a single input should be an error; feel free to define a better error type ("duplicate datom") if one doesn't exist.

eoger commented:
Updated, thanks.
I removed most of the code that was in that `for term in final_terms` loop since #533 does the duplicate-detection heavy-lifting now.

rnewman commented on db/src/tx.rs:
```diff
@@ -633,13 +654,16 @@ impl<'conn, 'a> Tx<'conn, 'a> {
-                          self.schema.require_attribute_for_entid(entids::DB_TX_INSTANT).unwrap(),
-                          TypedValue::Instant(self.tx_instant),
-                          true));
+        tx_instant = self.tx_instant.unwrap_or(::now());
```
`unwrap_or_else(TypeValue::current_instant)`.

rnewman commented on db/src/tx.rs:
```diff
@@ -633,13 +654,16 @@ impl<'conn, 'a> Tx<'conn, 'a> {
+        tx_instant = self.tx_instant.unwrap_or(::now());
+
+        // Transact [:db/add :db/txInstant NOW :db/tx] if it doesn't exist.
+        if self.tx_instant == None {
```
You always set `tx_instant` right above this. I don't think you need this conditional check (or the comment).

rnewman replied:
Yeah, but the point is: we've filtered out all txInstant-setting datoms from the input. Every tx is supposed to have one. So `tx_instant` (the local var) is the value we want, and we should always add it to the output. Right?

rnewman replied:
Oh, you stripped out the `continue`. Got it.

rnewman commented:
3bf74593
